File: Models/joining request models/coach_to_club_request.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class CoachToClubRequest {
  final String id;
  final String coachId;
  final String coachName;
  final String clubId;
  final String clubName;
  final String message;
  final String status;
  final DateTime requestedAt;
  final String? reviewedBy;
  final DateTime? reviewedAt;
  final String? reviewNote;
  final DateTime createdAt;
  final DateTime updatedAt;

  const CoachToClubRequest({
    required this.id,
    required this.coachId,
    required this.coachName,
    required this.clubId,
    required this.clubName,
    this.message = '',
    this.status = 'pending',
    required this.requestedAt,
    this.reviewedBy,
    this.reviewedAt,
    this.reviewNote,
    required this.createdAt,
    required this.updatedAt,
  });

  factory CoachToClubRequest.create({
    required String id,
    required String coachId,
    required String coachName,
    required String clubId,
    required String clubName,
    String? message,
  }) {
    final now = DateTime.now();
    return CoachToClubRequest(
      id: id,
      coachId: coachId,
      coachName: coachName,
      clubId: clubId,
      clubName: clubName,
      message: message ?? '',
      requestedAt: now,
      createdAt: now,
      updatedAt: now,
    );
  }

  Map<String, dynamic> toFirestore() => {
    'id': id,
    'coachId': coachId,
    'coachName': coachName,
    'clubId': clubId,
    'clubName': clubName,
    'message': message,
    'status': status,
    'requestedAt': Timestamp.fromDate(requestedAt),
    'reviewedBy': reviewedBy,
    'reviewedAt': reviewedAt == null ? null : Timestamp.fromDate(reviewedAt!),
    'reviewNote': reviewNote,
    'createdAt': Timestamp.fromDate(createdAt),
    'updatedAt': Timestamp.fromDate(updatedAt),
  };

  factory CoachToClubRequest.fromFirestore(
    Map<String, dynamic> json,
  ) => CoachToClubRequest(
    id: json['id'] ?? '',
    coachId: json['coachId'] ?? '',
    coachName: json['coachName'] ?? '',
    clubId: json['clubId'] ?? '',
    clubName: json['clubName'] ?? '',
    message: json['message'] ?? '',
    status: json['status'] ?? 'pending',
    requestedAt:
        (json['requestedAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    reviewedBy: json['reviewedBy'],
    reviewedAt: (json['reviewedAt'] as Timestamp?)?.toDate(),
    reviewNote: json['reviewNote'],
    createdAt: (json['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    updatedAt: (json['updatedAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
  );

  CoachToClubRequest copyWith({
    String? id,
    String? coachId,
    String? coachName,
    String? clubId,
    String? clubName,
    String? message,
    String? status,
    DateTime? requestedAt,
    String? reviewedBy,
    DateTime? reviewedAt,
    String? reviewNote,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) => CoachToClubRequest(
    id: id ?? this.id,
    coachId: coachId ?? this.coachId,
    coachName: coachName ?? this.coachName,
    clubId: clubId ?? this.clubId,
    clubName: clubName ?? this.clubName,
    message: message ?? this.message,
    status: status ?? this.status,
    requestedAt: requestedAt ?? this.requestedAt,
    reviewedBy: reviewedBy ?? this.reviewedBy,
    reviewedAt: reviewedAt ?? this.reviewedAt,
    reviewNote: reviewNote ?? this.reviewNote,
    createdAt: createdAt ?? this.createdAt,
    updatedAt: updatedAt ?? this.updatedAt,
  );

  bool get isPending => status == 'pending';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';
}

File: src/model/coach_to_club_request.ts
import { Timestamp } from 'firebase/firestore';

// Club Model
export interface ClubModel {
  id: string;
  name: string;
  location: string;
  description: string;
  contactEmail: string;
  isActive: boolean;
  totalCoaches: number;
  totalPupils: number;
  createdAt: Date;
  updatedAt: Date;
}

export const clubModelFromFirestore = (data: any, id: string): ClubModel => {
  const toDateTime = (v: any): Date => {
    if (v instanceof Timestamp) return v.toDate();
    return new Date();
  };

  return {
    id,
    name: data.name || '',
    location: data.location || '',
    description: data.description || '',
    contactEmail: data.contactEmail || '',
    isActive: data.isActive ?? true,
    totalCoaches: data.totalCoaches || 0,
    totalPupils: data.totalPupils || 0,
    createdAt: toDateTime(data.createdAt),
    updatedAt: toDateTime(data.updatedAt),
  };
};

export const clubModelToFirestore = (club: ClubModel): any => {
  return {
    name: club.name,
    location: club.location,
    description: club.description,
    contactEmail: club.contactEmail,
    isActive: club.isActive,
    totalCoaches: club.totalCoaches,
    totalPupils: club.totalPupils,
    createdAt: Timestamp.fromDate(club.createdAt),
    updatedAt: Timestamp.fromDate(club.updatedAt),
  };
};

// Coach to Club Request Model
export interface CoachToClubRequest {
  id?: string;
  coachId: string;
  coachName: string;
  clubId: string;
  clubName: string;
  message: string;
  status: 'pending' | 'approved' | 'rejected';
  requestedAt: Date;
  reviewedBy?: string;
  reviewedAt?: Date;
  reviewNote?: string;
  createdAt: Date;
  updatedAt: Date;
}

export const coachToClubRequestFromFirestore = (data: any): CoachToClubRequest => {
  return {
    id: data.id || '',
    coachId: data.coachId || '',
    coachName: data.coachName || '',
    clubId: data.clubId || '',
    clubName: data.clubName || '',
    message: data.message || '',
    status: data.status || 'pending',
    requestedAt: data.requestedAt instanceof Timestamp ? data.requestedAt.toDate() : new Date(),
    reviewedBy: data.reviewedBy,
    reviewedAt: data.reviewedAt instanceof Timestamp ? data.reviewedAt.toDate() : undefined,
    reviewNote: data.reviewNote,
    createdAt: data.createdAt instanceof Timestamp ? data.createdAt.toDate() : new Date(),
    updatedAt: data.updatedAt instanceof Timestamp ? data.updatedAt.toDate() : new Date(),
  };
};

export const coachToClubRequestToFirestore = (request: CoachToClubRequest): any => {
  return {
    id: request.id,
    coachId: request.coachId,
    coachName: request.coachName,
    clubId: request.clubId,
    clubName: request.clubName,
    message: request.message,
    status: request.status,
    requestedAt: Timestamp.fromDate(request.requestedAt),
    reviewedBy: request.reviewedBy || null,
    reviewedAt: request.reviewedAt ? Timestamp.fromDate(request.reviewedAt) : null,
    reviewNote: request.reviewNote || null,
    createdAt: Timestamp.fromDate(request.createdAt),
    updatedAt: Timestamp.fromDate(request.updatedAt),
  };
};